components/provider_ppucd/src/parse_ppucd.rs

```diff
@@ -0,0 +1,431 @@
+    blocks: &HashMap<UnicodeSet, HashMap<String, String>>,
+    defaults: &HashMap<String, String>,
+) -> HashMap<String, String> {
+    // create the map of applicable property values, and initialize it to
+    // defaults for the whole Unicode range
+    let mut prop_vals: HashMap<String, String> = HashMap::new();
+    prop_vals.clone_from(defaults);
+
+    // determine if this code point matches any of the block-wide defaults
+    // ("overrides")
+    for (range, block_prop_vals) in blocks {
+        if range.contains(char::from_u32(code_point).unwrap()) {
+            let block_prop_vals_clone = block_prop_vals.iter().map(|(k, v)| (k.clone(), v.clone()));
+            prop_vals.extend(block_prop_vals_clone);
+        }
+    }
+
+    // finally, apply the overrides for this code point
+    for (range, code_point_prop_vals) in code_point_overrides {
+        if range.contains(char::from_u32(code_point).unwrap()) {
+            let code_point_override_prop_vals_clone = code_point_prop_vals
+                .iter()
+                .map(|(k, v)| (k.clone(), v.clone()));
+            prop_vals.extend(code_point_override_prop_vals_clone);
+        }
+    }
+
+    prop_vals
+}
+
+fn get_binary_prop_unisets(
+    prop_aliases: &HashMap<String, HashSet<String>>,
+    code_points: &HashMap<u32, HashMap<String, String>>,
+) -> HashMap<String, UnicodeSet> {
+    let mut m: HashMap<String, UnicodeSet> = HashMap::new();
+
+    for (canonical_name, all_names) in prop_aliases {
+        let mut uniset_builder = UnicodeSetBuilder::new();
+        for (code_point, code_point_prop_vals) in code_points {
+            let code_point_prop_names: HashSet<String> =
+                code_point_prop_vals.keys().cloned().collect();
+            if !all_names.is_disjoint(&code_point_prop_names) {
+                uniset_builder.add_char(char::from_u32(*code_point).unwrap());
+            }
+        }
+        if !&uniset_builder.is_empty() {
+            let uniset = uniset_builder.build();
+            m.insert(canonical_name.clone(), uniset);
+        }
+    }
+
+    m
+}
+
+pub fn parse(s: String) -> UnicodeProperties {
+    let lines: std::str::Lines = s.lines();
+
+    let parseable_lines = lines.filter(|line| !is_skip_ppucd_line(line));
+
+    let mut prop_aliases: HashMap<String, HashSet<String>> = HashMap::new();
+
+    let mut defaults: HashMap<String, String> = HashMap::new();
+
+    // UnicodeSet is used to store the code point range described in a PPUCD
+    // `block` line
+    let mut blocks: HashMap<UnicodeSet, HashMap<String, String>> = HashMap::new();
+
+    // UnicodeSet is used to store the code point or code point range described
+    // in a PPUCD `cp` line, according to the PPUCD file format spec.
+    let mut code_point_overrides: HashMap<UnicodeSet, HashMap<String, String>> = HashMap::new();
+
+    let mut code_points: HashMap<u32, HashMap<String, String>> = HashMap::new();
+
+    for line in parseable_lines {
```

Comment:
It *appears* that you are walking forward through this file, line by line.  You never walk backward or perform any other type of random access.  Therefore, it would be more efficient, perhaps substantially so, if you didn't have to stick the whole file into memory before parsing it.  Read it line by line from a `io::Read` instead.

Reply:
Yeah, I was thinking the same thing. I created an issue to represent this optimization sub-task as #454 in case you are okay with it not blocking this particular PR, with the understanding that is still a part of the overall work needed to be done here.